Promotion of Topics and Schemas

aindriu-aiven opened the pull request:
About this change - What it does
This change adds documentation on how promotion of Topics and Schemas work, including server configuration required to be set up by the superadmin.

This change also adds some clarity as to where to find the clusterid where to use it and also adds it into a couple of document pages where the information was missing.

Why this way
The hope is to provide clear concise documentation that will help users quickly integrate klaw with their organisation.
Signed-off-by: Aindriu Lavelle <aindriu.lavelle@aiven.io>

muralibasani commented on docs/docs/howto/clusterconnectivity/aiven-kafka-cluster-sasl-ssl-protocol.rst:
```diff
@@ -6,7 +6,7 @@ This section provides you with information on how to connect Aiven for Apache Ka
 ------------
 
-* Set up the connection between the Klaw APIs (Core API and Cluster API), see :doc:`klaw-core-with-clusterapi`. This involves configuring the ``klaw.clusterapi.url`` setting in the Klaw UI and testing the connectivity to ensure the two APIs can communicate.
+* Set up the connection between the Klaw APIs (Core API and Cluster API) to use secure SSL, see :doc:`klaw-core-with-clusterapi`. This involves configuring the ``klaw.clusterapi.url`` setting in the Klaw UI and testing the connectivity to ensure the two APIs can communicate over https.
```
"to use secure SSL" it can also be without ssl right ?

aindriu-aiven replied:
My apologies I think you are right, I tested "Connect with Aiven for Apache Kafka® Cluster using SSL protocol" with http and it wouldn't work until i upgraded to SSL but with SASL it might be different I will update that.

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+=========
+
+
+Klaw allows users to control the creation of resources in their Kafka environment through a concept of promotion.
```
```suggestion
Klaw allows users to control the creation of resources in their Kafka environment through the concept of promotion.
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+- By maintaining a history of applied changes, easy track back on the evolution of a configuration.
+
+
+Promotion of resources are administered by the resource owners, each team has the ability to promote a topic, schema or connector from one environment to the next.
```
```suggestion
The resource owners administer promotion of resources, and each team can promote a topic, schema or connector from one environment to the next.
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+Configure environment promotion
+-------------------------------
+    Server Configuration
+      Found under **Dashboard -> Settings** the Superadmin user has the ability to define the ``Tenant Model`` which configures which environments promote to the next environment in your organisation.
```
```suggestion
      Found under **Dashboard -> Settings**, the Superadmin user has the ability to define the ``Tenant Model``, which configures which environments promote to the next environment in your organization.
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+      Found under **Dashboard -> Settings** the Superadmin user has the ability to define the ``Tenant Model`` which configures which environments promote to the next environment in your organisation.
+      Each resource Topic/Schema must already be created by the superadmin under ``Environments`` before being added to the ``Tenant Model`` or the server will reject the configuration.
+      Below in the example we can see that the Kafka Topic environments have been defined as 'Dev' & 'TST' the order is also specified which is the order that will be enforced in the promotion of topics.
+      Similarily the Schema Registry environments have all been defined and placed in order.
```
```suggestion
      Similarly, the Schema Registry environments have all been defined and placed in order.
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+
+    Topic Promotion
+      Once a topic is created in the base environment it is then possible to promote this to the next environment.
+      This will create a promotion request that can be reviewed and approved or declined by the requesters team mates. Each created environment is able to be seen in the same ``Topic Overview`` where the topic can be promoted.
```
```suggestion
      This will create a promotion request that can be reviewed, approved, or declined by the requester's teammates. Each created environment can be seen in the same ``Topic Overview`` where the topic can be promoted.
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+      Once a topic is created in the base environment it is then possible to promote this to the next environment.
+      This will create a promotion request that can be reviewed and approved or declined by the requesters team mates. Each created environment is able to be seen in the same ``Topic Overview`` where the topic can be promoted.
+
+        How to Promote a Topic
```
```suggestion
        How to Promote a Topic:
        
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+        6. A Topic request is now created for team members to review and approve under Approvals.
+
+    Schema Promotion
+      Under the ``Topic Overview`` the schema can be requested and viewed for an individual topic and as of Release 2.0.0 the ability to promote existing schemas to higher level environments is also available.
```
```suggestion
      Under the ``Topic Overview``, the schema can be requested and viewed for an individual topic. As of Release 2.0.0, the ability to promote existing schemas to higher-level environments is also available. The requester can select a particular version of the schema to promote from the lower environment to the higher environment.
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+      The requester can select a particular version of the schema to promote from the lower environment to the higher environment.
+
+
+        How to Promote a Schema
```
```suggestion
        How to Promote a Schema:
        
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+        3. Select the **Schema** tab under the main Topic section.
+        4. A button is available to promote a schema to a higher environment where a higher environment has been configured. ``Promote to [Next Environment]``
+        5. Select the version of the Schema you wish to promote to the higher environment, this Schema will be available for the team to review when approving or declining the request.
+        6. Optionally if the Schema you wish to promote is not compatible with the existing schemas on that topic ``Force Register Schema`` can be used to register the Schema.
```
```suggestion
        6. Optionally, if the Schema you wish to promote is not compatible with the existing schemas on that topic, ``Force Register Schema`` can be used to register the Schema.
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+        8. A schema request is now created for team members to review and approve under Approvals.
+
+    How does Force Register work?
+        The ``Force Register Schema`` option, for use when a schema needs to be promoted to the higher environment but which does not meet the compatibility of the previous schemas.
```
```suggestion
        When promoting a schema to a higher environment, the ``Force Register`` Schema option allows registration even if it is not compatible with previous schemas.
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+
+    How does Force Register work?
+        The ``Force Register Schema`` option, for use when a schema needs to be promoted to the higher environment but which does not meet the compatibility of the previous schemas.
+        If this option is selected, then klaw will change the compatibility of the topic to ``NONE``, register the new schema and then revert to the previous topic compatibility.
```
```suggestion
        If this option is selected, then Klaw will change the compatibility of the topic to ``NONE``, register the new schema and then revert to the previous topic compatibility.
```

harshini-rangaswamy commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,72 @@
+    How does Force Register work?
+        The ``Force Register Schema`` option, for use when a schema needs to be promoted to the higher environment but which does not meet the compatibility of the previous schemas.
+        If this option is selected, then klaw will change the compatibility of the topic to ``NONE``, register the new schema and then revert to the previous topic compatibility.
+        If the topic compatibility is not set it will fall back to the global compatibility however Klaw will not change the global compatibility.
```
```suggestion
        If the topic compatibility is not set, it will fall back to the global compatibility. However, Klaw will not change the global compatibility.
```

harshini-rangaswamy reviewed (changes requested):
I made some minor text suggestions.  Added line breaks.

aindriu-aiven commented:
> I made some minor text suggestions. Added line breaks.

Thanks I think I have added all the updates that you recommended and one from Murali as well.

muralibasani commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,74 @@
+Promotion
```
I think this file should be included in concepts.rst

muralibasani commented on docs/docs/concepts/promotion.rst:
````diff
@@ -0,0 +1,74 @@
+}
+
+```
+
````
Few extra line spaces ?

muralibasani commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,74 @@
+
+    Topic Promotion
+      Once a topic is created in the base environment it is then possible to promote this to the next environment.
+      This will create a promotion request that can be reviewed, approved, or declined by the requester's teammates. Each created environment can be seen in the same ``Topic Overview`` where the topic can be promoted.
```
this sounds a bit confusing 'Each created environment'

muralibasani reviewed (commented):
Few minor suggestions

muralibasani commented on docs/docs/concepts/promotion.rst:
```diff
@@ -0,0 +1,74 @@
+
+    How does Force Register work?
+        When promoting a schema to a higher environment, the ``Force Register`` Schema option allows registration even if it is not compatible with previous schemas.
+        If this option is selected, then Klaw will change the compatibility of the topic to ``NONE``, register the new schema and then revert to the previous topic compatibility.
```
Instead of topic compatibility, we can rename to 'subject compatibility'

muralibasani commented on docs/docs/howto/clusterconnectivity/aiven-kafka-cluster-sasl-ssl-protocol.rst:
```diff
@@ -30,7 +30,7 @@ Follow the steps below to configure and connect to an Aiven for Apache Kafka® a
 6. Click **Add Environment** and enter the details to add your Kafka environment. 
 7. Enter an environment name, set the cluster you added from the drop-down list, and configure partitions and replication factor, and tenat (set to default). 
-8. Copy the **Cluster ID** from the **Clusters** page using the copy icon.
+8. Copy the **Cluster ID** from the **Clusters** page using the copy icon that is available on the right hand side of the each cluster row.
```
this is good.

muralibasani commented on docs/docs/howto/clusterconnectivity/aiven-kafka-cluster-ssl-protocol.rst:
```diff
@@ -7,7 +7,7 @@ This section provides you with information on how to connect Aiven for Apache Ka
 ------------
 
-* Set up the connection between the Klaw APIs (Core API and Cluster API), see :doc:`klaw-core-with-clusterapi`. This involves configuring the ``klaw.clusterapi.url`` setting in the Klaw UI and testing the connectivity to ensure the two APIs can communicate.
+* Set up the connection between the Klaw APIs (Core API and Cluster API) to use secure SSL, see :doc:`klaw-core-with-clusterapi`. This involves configuring the ``klaw.clusterapi.url`` setting in the Klaw UI and testing the connectivity to ensure the two APIs can communicate over https.
```
As kafka ssl connectivity does not use application's security context, this should not be a pre.

muralibasani commented on docs/docs/howto/clusterconnectivity/aiven-kafka-connect-cluster-ssl-protocol.rst:
```diff
@@ -26,8 +29,9 @@ Follow the steps below to configure and connect an Aiven for Apache Kafka Connec
 6. In the **Kafka Connect Environments** section, click **Add Environment** and enter the details to add your schema registry environment. Click **Save**. 
-7. Open the ``application.properties`` file for `core`` (klaw/core/src/main/resources) and `cluster-api` (klaw/cluster-api/src/main/resources) modules. 
-8. Configure the SSL properties to connect to Schema Registry cluster by copying and editing the following lines:
+7. Open the ``application.properties`` file for `core`` (klaw/core/src/main/resources) and `cluster-api` (klaw/cluster-api/src/main/resources) modules.
```
May be the application properties only from cluster-api is needed here

aindriu-aiven replied:
The SSL properties if configured on the cluster api must be configured on core or it won't be able to communicate over https to the cluster api.

muralibasani commented on docs/docs/howto/clusterconnectivity/aiven-karapace-cluster-ssl-protocol.rst:
```diff
@@ -26,8 +26,9 @@ Follow the steps below to configure and connect to a Karapace schema registry cl
 6. In the **Schema Registry Environments** section, click **Add Environment** and enter the details to add your schema registry environment. Click **Save**. 
-7. Open the ``application.properties`` file for `core`` (klaw/core/src/main/resources) and `cluster-api` (klaw/cluster-api/src/main/resources) modules. 
-8. Configure the SSL properties to connect to Schema Registry cluster by copying and editing the following lines. 
+7. Open the ``application.properties`` file for `core`` (klaw/core/src/main/resources) and `cluster-api` (klaw/cluster-api/src/main/resources) modules.
```
also here, we would need only cluster-api properties file to be edited

muralibasani commented on docs/docs/howto/clusterconnectivity/kafka-cluster-ssl-protocol.rst:
```diff
@@ -7,7 +7,7 @@ This section provides you with information on how to connect Apache Kafka cluste
 Prerequisite
 ------------
-* Set up the connection between the Klaw APIs (Core API and Cluster API), see :doc:`klaw-core-with-clusterapi`. This involves configuring the ``klaw.clusterapi.url`` setting in the Klaw UI and testing the connectivity to ensure the two APIs can communicate.
+* Set up the connection between the Klaw APIs (Core API and Cluster API) to use secure SSL, see :doc:`klaw-core-with-clusterapi`. This involves configuring the ``klaw.clusterapi.url`` setting in the Klaw UI and testing the connectivity to ensure the two APIs can communicate over https.
```
also here, as kafka ssl connectivity does not use application's security context, this should not be a pre.
